### ISSUE TITLE
StackOverflow with v2.0.0

### ISSUE THREAD

**Issue by Mike-E-angelo** [seesharper/LightInject.Microsoft.Hosting#33]
Heya @seesharper 😁👋  ... I unfortunately appear to run into a StackOverflowException when upgrading this library to `2.0.0`, and I am thinking it is due to the new work that I am responsible for kicking into existence. 😇

More unfortunately, it doesn't give a lot of details, only that it seems to occur when I add Azure WebJobs.  The repeated stack is here:
https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Hosting/WebJobsServiceCollectionExtensions.cs#L79

Stacktrace here for your review:
![image](https://github.com/user-attachments/assets/59cc38de-700d-4628-b9c7-efcb2f73c4d8)

Reverting back to 1.6.1 avoids this issue.

Please let me know if you require any further information (quite possibly and understandable 😀) and I will assist you.

**Comment by seesharper** [comment]
Hi @Mike-E-angelo 

I need to investigate this, but what comes to mind here is variance. 
Unfortunately this is a little bit cumbersome to turn off at the moment.
If you have a `ICompositionRoot` somewhere could you try the following? 

```c#
 var optionsField = typeof(ServiceContainer).GetField("options", BindingFlags.Instance | BindingFlags.NonPublic);
 var options = (ContainerOptions)optionsField!.GetValue(registry);
 options.EnableVariance = false;
```

**Comment by Mike-E-angelo** [comment]
That did it @seesharper 👍  Turns out I create the container as part of my composition:
https://github.com/DragonSpark/Framework/blob/ef7d1fd5bc05f3bab81e5e51392a5cc628a36b24/DragonSpark.Composition/Construction/WithComposition.cs#L28

Is this a new field?  Anything to be aware of by setting it to false?

**Comment by seesharper** [comment]
The property is not new, but somehow I've managed to flip the default somewhere during the process here. 😩
This should not affect anything unless you specifically rely on variance in your code. 
Ih short, variance means that when resolving an `IEnumerable<T>`, LightInject will bring back all services compatible with the requested service type. MS.EX.DI does not support this as it is a pretty advanced feature. 

There is also a `VarianceFilter` property which can be used in situations where variance is required. By applying a filter it is possible to specify for which types to apply variance. For instance matching a given namespace. 

I will create an overload for the `UseLightInject` method that passes a func with the `ContainerOptions` so that we can get rid of that dirty reflection trick 👍😃

**Comment by Mike-E-angelo** [comment]
> The property is not new, but somehow I've managed to flip the default somewhere during the process here. 😩

Hah!  OK that sounds like we found the "root" cause then 😀

> I will create an overload for the UseLightInject method that passes a func with the ContainerOptions so that we can get rid of that dirty reflection trick 👍😃

That would be great!  Thank you for your continued efforts. 🙏

**Comment by seesharper** [comment]
Closed by #34
